lib/node_modules/@stdlib/array/base/slice/docs/types/index.d.ts

```diff
@@ -37,204 +37,16 @@ import { Collection, Complex128Array, Complex64Array } from '@stdlib/types/array
-* @returns output array
-*
-* @example
-* var Uint16Array = require( '@stdlib/array/uint16' );
-*
-* var x = new Uint16Array( [ 1, 2, 3 ] );
-*
-* var out = slice( x, 0, 3 );
-* // returns <Uint16Array>[ 1, 2, 3 ]
-*/
-declare function slice( x: Uint16Array, start: number, end: number ): Uint16Array;
-
-/**
-* Returns a shallow copy of a portion of an array.
-*
-* @param x - input array
-* @param start - starting index (inclusive)
-* @param end - ending index (exclusive)
-* @returns output array
-*
-* @example
-* var Uint8Array = require( '@stdlib/array/uint8' );
-*
-* var x = new Uint8Array( [ 1, 2, 3 ] );
-*
-* var out = slice( x, 0, 3 );
-* // returns <Uint8Array>[ 1, 2, 3 ]
-*/
-declare function slice( x: Uint8Array, start: number, end: number ): Uint8Array;
-
-/**
-* Returns a shallow copy of a portion of an array.
-*
-* @param x - input array
-* @param start - starting index (inclusive)
-* @param end - ending index (exclusive)
-* @returns output array
-*
-* @example
-* var Uint8ClampedArray = require( '@stdlib/array/uint8c' );
-*
-* var x = new Uint8ClampedArray( [ 1, 2, 3 ] );
-*
-* var out = slice( x, 0, 3 );
-* // returns <Uint8ClampedArray>[ 1, 2, 3 ]
-*/
-declare function slice( x: Uint8ClampedArray, start: number, end: number ): Uint8ClampedArray;
-
-/**
-* Returns a shallow copy of a portion of an array.
-*
-* @param x - input array
-* @param start - starting index (inclusive)
-* @param end - ending index (exclusive)
-* @returns output array
-*
-* @example
-* var Complex128Array = require( '@stdlib/array/complex128' );
-*
-* var x = new Complex128Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
-*
-* var out = slice( x, 0, 3 );
-* // returns <Complex128Array>[ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ]
-*/
-declare function slice( x: Complex128Array, start: number, end: number ): Complex128Array;
-
-/**
-* Returns a shallow copy of a portion of an array.
-*
-* @param x - input array
-* @param start - starting index (inclusive)
-* @param end - ending index (exclusive)
-* @returns output array
 *
 * @example
 * var Complex64Array = require( '@stdlib/array/complex64' );
 *
-* var x = new Complex64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
-*
-* var out = slice( x, 0, 3 );
-* // returns <Complex64Array>[ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ]
-*/
-declare function slice( x: Complex64Array, start: number, end: number ): Complex64Array;
-
-/**
-* Returns a shallow copy of a portion of an array.
-*
-* @param x - input array
-* @param start - starting index (inclusive)
-* @param end - ending index (exclusive)
-* @returns output array
-*
-* @example
-* var x = [ 1, 2, 3, 4, 5, 6 ];
+* var x = new Complex64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
 *
 * var out = slice( x, 0, 3 );
-* // returns [ 1, 2, 3 ]
+* // returns <Complex64Array>[ 1.0, 2.0, 3.0 ]
 */
-declare function slice<T = unknown>( x: Collection<T>, start: number, end: number ): Array<T>;
+declare function slice<T extends Collection | ComplexTypedArray>( x: T, start: number, end: number ): T;
```

Comment:
I think we lost some info here. Namely, if `T` is a collection without a `slice` method, we always return a "generic" array having elements of the same type. The prior definition was a fall back definition for "unknown" collections and "generic" arrays. We can probably do better by defining an interface which extends `Collection` and has a `slice` method conforming to a particular signature. In that case, `T` in and `T` out. Otherwise, for a collection without a `slice` method, we return a "generic" array having elements of type `U`.

Reply:
In my mind, that the function does not apply to strings is fine with me and part of the original reason why I typed it as `Collection`, rather than `ArrayLike`. For users wanting to slice strings, they'd be better off using a dedicated `@stdlib/string/slice` package.

Reply:
@kgryte Makes sense, although I believe `strings` do fall under our `Collection`s as well, right?

Reply:
Actually, no.

See https://github.com/stdlib-js/stdlib/blob/4315e33177a67ff0ca097e7dcf91c40f90380e91/tools/docs/jsdoc/typedefs/objects.js#L54 and https://github.com/stdlib-js/stdlib/blob/4315e33177a67ff0ca097e7dcf91c40f90380e91/lib/node_modules/%40stdlib/assert/is-collection/lib/main.js#L45.

Reply:
Strings do fall under the TypeScript definition (https://github.com/stdlib-js/stdlib/blob/4315e33177a67ff0ca097e7dcf91c40f90380e91/lib/node_modules/%40stdlib/types/index.d.ts#L595); but this was because we, I believe, could not easily type an "array-like object".

Reply:
Things may have changed, but IIRC it may have been influenced by TypeScript 2.1 capabilities.